_posts/2022-09-20-internship-projects-2022-concrete-playback.md

```diff
@@ -0,0 +1,239 @@
+    let concrete_vals: Vec<Vec<u8>> = vec![
+        // 0ul
+        vec![0, 0, 0, 0, 0, 0, 0, 0],
+        // 301283365829122ul
+        vec![2, 58, 255, 255, 3, 18, 1, 0],
+        // 36573ul
+        vec![221, 142, 0, 0, 0, 0, 0, 0]
+    ];
+    kani::concrete_playback_run(concrete_vals, stretched_rectangle_can_hold_original);
+}
+```
+INFO: To automatically add the concrete playback unit test `kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517` to the src code, run Kani with `--concrete-playback=InPlace`.
+````
+
+With the `print` mode, concrete playback generates a Rust unit test that:
+ 1. Initializes a vector of byte vectors with concrete values.
+ 2. Runs the harness we tested with those concrete values.
+
+The way the vector `concrete_vals` is initialized may be confusing at first.
+But if you look closely, you'll realize that each vector contained in the main
+vector is a sequence of eight bytes (or `u8` values), which is just another way
+to represent a `u64` value. In summary, Kani requires each variable assigned
+`kani::any()` in our harness (`factor`, `width`, and `height`) to be initialized
+byte by byte, and that's why we need three 8-byte vectors.
+Because bytes are not easy to read for humans, we add a comment above each bytes
+vector with the value it represents.
+
+Now let's run Kani with the `inplace` mode:
+
+```bash
+cargo kani --harness stretched_rectangle_can_hold_original --enable-unstable --concrete-playback=inplace
+# --snip--
+INFO: Parsing concrete values from property `rectangle::stretched_rectangle_can_hold_original.assertion.1` with description `assertion failed: larger.can_hold(&original)`.
+INFO: Now modifying the source code to include the concrete playback unit test `kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517`.
+```
+
+The test unit that was printed out is now added to the source code
+automatically! Now we can run the test that Kani inserted into
+the source code, which can be done[^footnote-setup] with the default
+command for running Rust unit tests:
+
+```
+ cargo +nightly test
+```
+
+```
+running 1 test
+test rectangle::tests::kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517 ... FAILED
+
+failures:
+
+---- rectangle::tests::kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517 stdout ----
+thread 'rectangle::tests::kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517' panicked at 'assertion failed: larger.can_hold(&original)', src/rectangle.rs:36:9
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+
+
+failures:
+    rectangle::tests::kani_concrete_playback_stretched_rectangle_can_hold_original_11077055701284606517
+
+test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
+```
+
+There are some neat details about this mode that have implemented after watching
```

Comment:
sentence reads a little off. I believe it should read:

"There are some neat details about this mode that have _been_ implemented after watching..."

Reply:
Yes, this was likely a typo.